Feature/streaming by quality

Seokjin-K opened the pull request:
### 변경사항
<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->
**AS-IS**
- 오디오 품질을 변환할 기능 부재
- 현재 업로드 기능은 파일을 하나씩 업로드 함

**TO-BE**
- FFmpeg를 활용하여 오디오 품질 변환 기능 구현
- S3 저장소에 다중 파일 업로드 기능 구현

### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [ ] 테스트 코드
- [x] API 테스트

Seokjin-K committed:
feat: File 관련 Util 클래스 생성

- File 관련 Util 클래스 생성
- @Slf4j의 log와 multipartFile 사용을 위해 spring-boot-starter-web 의존성 추가

Seokjin-K committed:
feat: FFmpeg 사용을 위해 FFmpeg 의존성 추가

Seokjin-K committed:
rename: S3Config 경로 변경 및 S3StorageService에서 S3Stroage로 변경

Seokjin-K committed:
feat: FFmpeg를 사용한 음질 변환 파일 생성 구현

- AudioQualit 파라미터를 기반으로 해당 음질에 맞게 파일 재생성
- FFmpeg 설정을 위한 FFmpegConfig 구현

Seokjin-K committed:
feat: 여러 음질 보관을 위해 FileKey 추가

Seokjin-K committed:
feat: 음원과 가사의 통합본을 표현하기 위한 Enum 속성 추가

Seokjin-K committed:
rename : 패키지 분할

- com.music.을 com.music.stream.과 com.music.upload.으로 분리

Seokjin-K committed:
remove: FileStorageService 이름 변경으로 인한 파일 제거

- FileStorageService에서 FileStorage로 이름 변경

Seokjin-K committed:
chore: yml 설정

- ffmpeg 관련 yml 설정
- mongoDB 관련 yml 설정

Seokjin-K committed:
feat: 파일 검증을 위한 FileValidator 구현

Seokjin-K committed:
feat: AudioConvert 인터페이스 생성

Seokjin-K committed:
rename: adaptor 패키지로 이동

Seokjin-K committed:
style: 사용하지 않는 코드 수정

Seokjin-K committed:
feat: 오디오 품질별 정보를 표현하기 위한 AudioQuality Enum 구현

Seokjin-K committed:
feat: 음원과 가사의 파일 키를 래핑하기 위한 FileKeys 클래스 추가

Seokjin-K committed:
feat: 음원과 가사를 업로드하기 위한 정보와 정보 제공 메서드를 포함한 클래스 추가

Seokjin-K committed:
feat: Enum 타입을 구분하여 각각 다른 validator 로직이 작동되도록 하는 FileValidatorFactory 구현

Seokjin-K committed:
feat: MultipartFile 형태의 데이터를 실제 파일로 변환하고 FFmpeg를 통해 품질을 변환하는 클래스 구현

- FFmpeg는 실제 파일을 기반으로 동작

Seokjin-K committed:
feat: s3에서 제공하는 ObjectMetadataProvider를 구현한 클래스 추가

- 해당 클래스를 통해 각 파일에 메타데이터를 세팅하는 Provider 제공

mingming-mentor commented on infrastructure/src/main/java/com/music/s3/CustomObjectMetadataProvider.java:
```diff
@@ -0,0 +1,62 @@
+      applyLyricsMetadata(file, metadata);
+      log.info("\napplyLyricsMetadata : {}\n", metadata);
+      log.info(
+          "\napplyLyricsMetadata for file {}: Content-Type: {}, Content-Length: {}, User Metadata: {}\n",
```
앞에 \n을 넣으신 이유가 있을까요? logger의 경우 println처럼 기본적으로 엔터를 붙입니다.

mingming-mentor commented on infrastructure/src/main/java/com/music/s3/FileValidator.java:
```diff
@@ -0,0 +1,63 @@
+import org.springframework.web.multipart.MultipartFile;
+
+@Slf4j
+@Component
```
bean으로 선언하신 이유가 있나요?

Seokjin-K replied:
사실 당시 기억이 잘 나지 않습니다. 아무래도 bean으로 선언하여 다른 곳에서 쉽게 주입받아 사용할 생각이었던 것 같은데, 이와 같은 이유라면 static 메서드로 선언하는 게 좋았을까요? 아니면 다른 접근 방식이 있을까요?

mingming-mentor commented on music/src/main/java/com/music/upload/dto/TrackUpload.java:
```diff
@@ -0,0 +1,38 @@
+  public List<File> getFiles() {
+    List<File> files = this.musicUpload.getFiles();
+    this.getLyricsUploadOptional().ifPresent(lyrics -> files.add(lyrics.getLyricsFile()));
+    System.out.println("\n\n------------------------- 가사 파일 추가 완료 -------------------------\n\n");
```
logger쓰시면 좋겠네요

Seokjin-K replied:
급하게 테스트 하느라 지우는 것을 잊어버렸네요. 수정하겠습니다.

mingming-mentor commented on music/src/main/java/com/music/upload/service/MusicUploadService.java:
```diff
@@ -0,0 +1,134 @@
+    Music music = createMusic(request, album, fileKeys);
+    musicRepository.save(music);
+
+    final String LYRICS_KEY = fileKeys.getLyricsKey();
```
snake case로 쓰신 이유가 있나요?

Seokjin-K replied:
final로 선언하여 습관적으로 snake case를 사용한 것 같은데, 생각해보니 메서드의 호출 결과를 저장하는 것이기 때문에 진정한 상수의 의미는 아닌 것 같네요. 수정하겠습니다.

mingming-mentor commented on music/src/main/java/com/music/upload/service/MusicUploadService.java:
```diff
@@ -0,0 +1,134 @@
+    musicRepository.save(music);
+
+    final String LYRICS_KEY = fileKeys.getLyricsKey();
+    if (LYRICS_KEY != null && !LYRICS_KEY.isEmpty()) {
+      Lyrics lyrics = Lyrics.builder()
+          .music(music)
+          .lyricsFileKey(fileKeys.getLyricsKey())
+          .lyricsFormat(LyricsMimeType.getFormatByContentType(lyricFile.getContentType()))
+          .build();
+
+      lyricsRepository.save(lyrics);
+    }
```
hasLyricsFile에서는 true가 나오는데 LYRICS_KEY가 null이거나 empty일수도 있나요? 
아니라면 같은 조건으로 체크하면 좋을 것 같네요

mingming-mentor commented on infrastructure/src/main/java/com/music/s3/S3Storage.java:
```diff
@@ -0,0 +1,148 @@
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      log.error("File upload interrupted : {}", e.getMessage());
+      throw new RuntimeException(); // TODO: CustomException 으로 변경
```
아래처럼 기본적으로 던져지는 e는 Exception에 같이 넣어서 넘겨주시면 좋을 것 같아요
```suggestion
      throw new RuntimeException(e); // TODO: CustomException 으로 변경
```

mingming-mentor reviewed (approved):
수고 많으셨습니다

Seokjin-K committed:
feat: 출력로그 개행문자 제거

Seokjin-K committed:
refactor: 필요하지 않은 코드 제거

Seokjin-K committed:
rename: 변수 LYRICS_KEY의 이름을 camel case로 변경

Seokjin-K committed:
feat: RuntimeException 생성자에 예외 정보 넘기도록 변경

Seokjin-K committed:
feat: 스트리밍할 때 해당 객체의 quality 메타데이터 출력 하도록 기능 추가

Seokjin-K committed:
feat: 퀄리티를 기준으로 파일키를 가져오도록 AudioQuality enum 기능 확장

- 퀄리티와 파일키를 가져오는 메서드를 매핑한 FILE_KEY_MAPPER 선언
- 해당 퀄리티의 파일키를 가져오는 getFIleKey 메서드 구현

Seokjin-K committed:
feat: 클라이언트에서 요청한 품질을 기준으로 스트리밍 제공하도록 변경
